### ISSUE TITLE
Considerations with path-to-regexp 2.0

### ISSUE THREAD

**Issue by pshrmn** [ReactTraining/react-router#5469]
`path-to-regexp` recently released `v2.0.0`. I thought that this should be discussed to determine if React Router should update.

The changelog is [here](https://github.com/pillarjs/path-to-regexp/blob/master/History.md#200--2017-08-23), but to save everyone a click, here are the changes:

> * New option! Ability to set endsWith to match paths like /test?query=string up to the query string
> * New option! Set delimiters for specific characters to be treated as parameter prefixes (e.g. /:test)
> * Remove isarray dependency
> * Explicitly handle trailing delimiters instead of trimming them (e.g. /test/ is now treated as /test/ instead of /test when matching)
> * Remove overloaded keys argument that accepted options
> * Remove keys list attached to the RegExp output
> * Remove asterisk functionality (it's a real pain to properly encode)
> * Change tokensToFunction (e.g. compile) to accept an encode function for pretty encoding (e.g. pass your own implementation)

#### New options

Since `<Route>`s limit the ability to access the options through props, these would not be available immediately. `path-to-regexp`'s documentation of the options is a bit frustrating, but you can look at the tests to get a better understanding.

`endsWith` should be unnecessary. Its use case seems to be matching pathnames that still have the search or hash segments included. Since we are matching against location objects that have already stripped these, there is no reason that I can immediately think of to use this.

`delimiters` could probably be supported, but I'm not sure that I've every seen anyone who wants their pathnames to be `$what$is$this$i$dont$even`. Supporting this would also possibly make partial routes more complicated (haven't given it much thought).

#### Explicit Trailing Delimiters

This should limit the need to use `strict`.

```js
// 1.7.0
<Route path='/test/' />
// ['/test', '/test/'] -> true, true

<Route path='/test/' strict />
// ['/test', '/test/'] -> false, true


// 2.0.0
<Route path='/test/' />
// ['/test', '/test/'] -> false, true
```

#### Keys

`key` handling has been changed, but this shouldn't actually have any effect here. Basically, the regular expression returned when calling `pathToRegexp` used have a `keys` property, but no longer does. `compilePath` never used this, so this change should not break anything.

#### Wildcards

The biggest change is that `*` is no longer valid. Instead, the recommendation is to use `(.*)`.

```js
// 1.7.0
<Route path='*' component={Wildcard} />
// 2.0.0
<Route path='(.*)' component={Wildcard} />
```

Removing wildcards means that the path is a bit uglier, but my understanding is that it made the matching RegExp simpler. Looking at the [diff](https://github.com/pillarjs/path-to-regexp/commit/9d4cbd91698529e719fc00d20dc87aa382db54ee#diff-168726dbe96b3ce427e7fedce31bb0bcR24), the new one is definitely shorter, but I don't feel like trying to mentally parse the differences between two complex regular expressions :man_shrugging:.

----

In summary, I think that updating to `path-to-regexp v2.0.0` should be a fairly painless. We could consider adding the `delimiters` prop to the `<Route>`, but I don't think that it is needed immediately.

There really aren't breaking changes, so I don't think that it warrants waiting until v5, but I think that it should at least be a minor release.

**Comment by pshrmn** [comment]
**Update** There are actually some issues with the explicit trailing delimiters. I'll need to do look into it some more, but this isn't working with `2.0.0`:
```js
<MemoryRouter initialEntries={[ '/nested' ]}>
  <Route path="/" render={() => (
    <div>
      <h1>{TEXT1}</h1>
      <Route path="/nested" render={() => (
        <h2>{TEXT2}</h2>
      )}/>
    </div>
  )}/>
</MemoryRouter>
```

**Comment by pshrmn** [comment]
Opened https://github.com/pillarjs/path-to-regexp/issues/116 to get feedback on the root matching issue.

**Comment by timdorr** [comment]
Note that this dropped just hours before I pushed out 4.2.0. I had quickly updated to it, but it hosed the test suite, so I kept it locked to 1.x. 

Also, to make things a little easier on everyone, here are the 1.x and 2.0 regexps with the double slashes stripped:
```
// 1.x
([\/.])?(?:(?:\:(\w+)(?:\(((?:\\.|[^\\()])+)\))?|\(((?:\\.|[^\\()])+)\))([+*?])?|(\*))

// 2.0
(?:\:(\w+)(?:\(((?:\\.|[^\\()])+)\))?|\(((?:\\.|[^\\()])+)\))([+*?])?
```

I recommend [Regex101](https://regex101.com/) for analysis and debugging. They give a good step-by-step view of what's going on in your expression.

**Comment by pshrmn** [comment]
I narrowed down the cause of the issue, so now we just need to wait and see.

When I stare at regular expressions for too long, it reminds me how much I look forward to [named capture groups](https://github.com/tc39/proposal-regexp-named-groups). `path-to-regexp` is great, but being able to use native syntax for path matching/param extraction should make life easier.

**Comment by pshrmn** [comment]
Okay, so something that has become clearer now is that `/` and `/test` should be considered siblings. `/` is essentially `/index` and it doesn't really make sense for it to have children.

Instead of treating a pathless `<Route>`'s `path` as `"/"`, we should use the empty string `""` to match everything.

I'll create a PR and the discussion can be moved to there.
